sdk/eventhub/azure-eventhubs/azure/eventhub/aio/consumer_async.py

```diff
@@ -186,45 +211,14 @@ def queue_size(self):
                 :caption: Receives events asynchronously
 
         """
-        max_batch_size = kwargs.get("max_batch_size", None)
-        timeout = kwargs.get("timeout", None)
-
         self._check_closed()
-        max_batch_size = min(self.client.config.max_batch_size, self.prefetch) if max_batch_size is None else max_batch_size
-        timeout = self.client.config.receive_timeout if timeout is None else timeout
-        if not timeout:
-            timeout = 100000  # timeout None or 0 mean no timeout. 100000 seconds is equivalent to no timeout
-
-        data_batch = []
-        start_time = time.time()
-        timeout_time = start_time + timeout
-        max_retries = self.client.config.max_retries
-        retry_count = 0
-        last_exception = None
-        while True:
-            try:
-                await self._open(timeout_time)
-                remaining_time = timeout_time - time.time()
-                if remaining_time <= 0.0:
-                    if last_exception:
-                        log.info("%r receive operation timed out. (%r)", self.name, last_exception)
-                        raise last_exception
-                    return data_batch
-
-                remaining_time_ms = 1000 * remaining_time
-                message_batch = await self._handler.receive_message_batch_async(
-                    max_batch_size=max_batch_size,
-                    timeout=remaining_time_ms)
-                for message in message_batch:
-                    event_data = EventData(message=message)
-                    self.offset = EventPosition(event_data.offset)
-                    data_batch.append(event_data)
-                return data_batch
-            except EventHubError:
-                raise
-            except Exception as exception:
-                last_exception = await self._handle_exception(exception, retry_count, max_retries, timeout_time)
-                retry_count += 1
+
+        timeout = timeout or self.client.config.receive_timeout
+        max_batch_size = max_batch_size or min(self.client.config.max_batch_size, self.prefetch)
+        data_batch = []  # type: List[EventData]
+
+        return await _retry_decorator(self._receive)(self, timeout=timeout,
```

Comment:
This is a somewhat unusual construction, which might surprise other people looking at the code expecting Python `@` decorator usage. It also means the function wrapper has to be created on every call, instead of just once, when the module is executed. I'd suggest using typical decorator syntax to make "retry" versions of these methods. e.g 
```
@_retry_decorator 
def _recieve_retryself, timeout):
    return setf.receive(timeout)
```
and then using those.

Reply:
Is this because the function name has "decorator" in it? I haven't seen many cases with a local function with a decorator like this either.

Reply:
Partly. This fits the broader definition of "decorator", but specifically in the context of Python, it has a strong association with `@foo` syntax.

Reply:
thanks @johanste @bryevdv , now I have a deeper understanding of the decorator.
I have applied what Bryan suggested, writing retry-version of **send/receive/open** methods. They are consistent now.